mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java

```diff
@@ -0,0 +1,282 @@
+        return InstanceHolder.INSTANCE;
+    }
+
+    public ConnectionFactory() {
+        if (Build.VERSION.SDK_INT >= 21) {
+            localConnectionTypes.add(ConnectivityManager.TYPE_VPN);
+        }
+    }
+
+    private void setContext(Context ctx) {
+        this.ctx = ctx;
+    }
+
+    @Override
+    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
+        if (needInvalidateCachePreferenceKeys.contains(key)) {
+            cachedConnections.clear();
+        }
+    }
+
+    private void setSettings(SharedPreferences settings) {
+        if (this.settings != null) {
+            this.settings.unregisterOnSharedPreferenceChangeListener(this);
+        }
+        this.settings = settings;
+
+        this.settings.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    public static Connection getConnection(int connectionType, Context ctx) {
+        return getConnection(connectionType, ctx,
+                PreferenceManager.getDefaultSharedPreferences(ctx));
+    }
+
+    public static Connection getConnection(int connectionType, Context ctx,
+                                           SharedPreferences settings) {
+        // FIXME: Once we're on API level 19, use Objects.requireNonNull();
+        if (ctx == null) {
+            throw new NullPointerException("Context ctx can not be null.");
+        }
+
+        ConnectionFactory factory = getInstance();
+        factory.setSettings(settings);
+        factory.setContext(ctx);
+
+        ctx.registerReceiver(factory, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
+
+        if (factory.cachedConnections.containsKey(Connection.TYPE_ANY)) {
+            return factory.cachedConnections.get(Connection.TYPE_ANY);
+        }
+
+        Connection conn;
+        switch (connectionType) {
+            case Connection.TYPE_LOCAL:
+                conn = factory.getLocalConnection();
+                break;
+            case Connection.TYPE_REMOTE:
+                conn = factory.getRemoteConnection();
+                break;
+            case Connection.TYPE_CLOUD:
+                // TODO: Need a proper way of finding if the connection supports openHAB cloud
+                // things, e.g. by checking if the /api/v1/settings/notifications endpoint works,
+                // but currently does not work for myopenhab.org
+                conn = factory.getRemoteConnection();
+                break;
+            case Connection.TYPE_ANY:
+                conn = factory.getAvailableConnection();
+                break;
+            default:
+                throw new IllegalArgumentException("Invalid Connection type requested.");
+        }
+
+        if (conn != null)
+            factory.cachedConnections.put(connectionType, conn);
+
+        return conn;
+    }
+
+    private Connection getLocalConnection() {
+        String openHABUrl = Util.normalizeUrl(settings.getString(Constants
+                .PREFERENCE_URL, ""));
```

Comment:
> Maybe out of scope, but the preference constants for url, user and pw are IMO hard to read. Can we change them to PREFERENCE_(REMOTE_)(URL|USERNAME|PASSWORD)?

> I agree! However, the PR already changes a lot, should I do it in this PR (would be pretty simple and not much lines to touch, I think). I'll do it in the next commit. If we think it's too much, I can still rever these changes :)

Reply:
Oops, changed and committed, will be here with the next push, sorry for missing that! :(